packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,888 @@
+    bucket._visitChildren(_recursivelyUpdateManager);
+  }
+
+  void _finalize() {
+    _childrenToAdd.forEach((RestorationId id, Set<RestorationBucket> buckets) {
+      assert(() {
+        final int claimCount = buckets.length + (_claimedChildren.containsKey(id) ? 1 : 0);
+        if (claimCount > 1) {
+          throw FlutterError.fromParts(<DiagnosticsNode>[
+            ErrorSummary('Multiple owners claimed child RestorationBuckets with the same ID.'),
+            ErrorDescription('The following owners claimed child RestorationBuckets with id "$id" from the parent $this:'),
+            ...buckets.map((RestorationBucket bucket) => ErrorDescription(
+              ' * ${bucket.debugOwner}',
+            )),
+            if (_claimedChildren.containsKey(id))
+              ErrorDescription(' * ${_claimedChildren[id].debugOwner} (current owner)'),
+          ]);
+        }
+        return true;
+      }());
+      _finalizeAddChildData(buckets.first);
+    });
+    _childrenToAdd.clear();
+  }
+
+  void _removeChildData(RestorationBucket child) {
+    assert(child != null);
+    assert(child._parent == this);
+    if (_claimedChildren.remove(child.id) == child) {
+      _rawChildren.remove(child.id.value);
+      if (_rawChildren.isEmpty) {
+        _rawData.remove(_childrenMapKey);
+      }
+      _manager?.scheduleSerialization();
+      return;
+    }
+    _childrenToAdd[child.id]?.remove(child);
+    if (_childrenToAdd[child.id]?.isEmpty == true) {
+      _childrenToAdd.remove(child.id);
+    }
+  }
+
+  void _addChildData(RestorationBucket child) {
+    assert(child != null);
+    assert(child._parent == this);
+    if (_claimedChildren.containsKey(child.id)) {
+      // Delay addition until the end of the frame in the hopes that the current
+      // owner of the child with the same id will have given up that child by
+      // then.
+      _childrenToAdd.putIfAbsent(child.id, () => <RestorationBucket>{}).add(child);
+      _manager.scheduleSerialization(finalizer: _finalize);
+      return;
+    }
+    _finalizeAddChildData(child);
+    _manager.scheduleSerialization();
+  }
+
+  void _finalizeAddChildData(RestorationBucket child) {
+    assert(_claimedChildren[child.id] == null);
+    assert(_rawChildren[child.id.value] == null);
+    _claimedChildren[child.id] = child;
+    _rawChildren[child.id.value] = child._rawData;
+  }
+
+  void _visitChildren(_BucketVisitor visitor, {bool concurrentModification = false}) {
+    Iterable<RestorationBucket> children = _claimedChildren.values
+        .followedBy(_childrenToAdd.values.expand((Set<RestorationBucket> set) => set));
+    if (concurrentModification) {
+      children = children.toList(growable: false);
+    }
+    children.forEach(visitor);
+  }
+
+  // Bucket management
+
+  /// Changes the [RestorationId] under which the bucket is stored in its parent
+  /// to `newId`.
+  ///
+  /// No-op if the bucket is already stored under the provided id.
+  void rename(RestorationId newId) {
```

Comment:
What happens if we claim the same name someone else already uses?

Can/should we assert that's not the case here?